### PR TITLE
🌱 Reuse capi upgrade test function

### PR DESCRIPTION
**Pull request by mboukhalfa**
This PR remove the custom upgrade test and reuse [clusterctl_upgrade](https://github.com/kubernetes-sigs/cluster-api/blob/release-1.2/test/e2e/clusterctl_upgrade.go) test from capi.

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
Will be removed, this is to test CI behaviour. 

/hold

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [inline test/e2e/upgrade_management_cluster_test.go:356]: we can add as a workaround to the path /home/metal3ci/.krew/bin

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [inline test/e2e/upgrade_management_cluster_test.go:54]: This needs a doc string and a better name. What is it doing?
Do you think it would be possible to define this as a variable instead before the `Describe`? I mean something like this:

```go
createBMHs := func(clusterProxy framework.ClusterProxy, clusterNamespace string) {...}
// ...
// Then use it like this
PreInit: createBMHs
```

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [inline test/e2e/upgrade_management_cluster_test.go:94]: Why is it set here instead of before we start the test? Is it because of the older version? In any case it needs an explanation.

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [inline test/e2e/upgrade_management_cluster_test.go:101]: Version is hard coded. We should parameterize this.

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [inline test/e2e/upgrade_management_cluster_test.go:133]: Same. This also needs a better name and doc string.

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [inline test/e2e/upgrade_management_cluster_test.go:69]: If we change the `splitBMHostFile` like suggested, we could make this easier to understand by looping over the file names directly. I would also want to use `clusterProxy.Apply` like for cert-manager below.

**Review by lentzi90 — CHANGES_REQUESTED** [review]
This will be nice when finished!
/hold
https://github.com/metal3-io/cluster-api-provider-metal3/pull/754 should go in first so we don't need the uplift in this one.

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:27]: I think here we are testing upgrade from `v1alpha4` (capi) so setting this in the environment somewhere then propagate it to here with some logic doesn't seem needed so far there is not case where `InitWithProvidersContract` will be `v1beta1` and since the support for this api versoin already dropped I do not think we will get patches for `v0.4.x` so `v0.4.7` is the latest.

It is hardcoded but it makes debugging the input easier although we can move them to e2e read them from there

File: scripts/ci-e2e.sh
Comment (furkatgofurov7) [inline scripts/ci-e2e.sh:61]: nit: typo

File: scripts/environment.sh
Comment (furkatgofurov7) [inline scripts/environment.sh:64]: is this needed?

File: scripts/environment.sh
Comment (furkatgofurov7) [inline scripts/environment.sh:67]: I am not understanding why we would need all of these

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:25]: can we make this a bit clearer, with TODO note, i.e:  `// TODO(mboukhalfa):  parameterize after merging https://github.com/kubernetes-sigs/cluster-api/pull/7373`

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:38]: with respect to Go naming conventions:
```suggestion
func installSplitYAML() {
```

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:50]: this has to be reworded definitely

File: test/e2e/config/e2e_conf.yaml
Comment (furkatgofurov7) [inline test/e2e/config/e2e_conf.yaml:152]: how is this specific version was chosen? based on `UPGRADED_K8S_VERSION`? do we need an extra variable at all?
Edit: I don't see it being used anywhere

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [reply test/e2e/upgrade_management_cluster_test.go:27]: @mboukhalfa you already have env var for them: `CAPI_FROM_RELEASE`

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:32]: parameterize would be better

**Review by furkatgofurov7 — COMMENTED** [review]
This looks like a nice improvement, however, I would like to understand better the workflow with these changes introduced:

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [reply test/e2e/upgrade_management_cluster_test.go:27]: Ok maybe v1alpha4 is fine but don't we already have the release here: https://github.com/metal3-io/cluster-api-provider-metal3/blob/6b3b651f81b54a48c9f39d5d402baf58c26af225/scripts/ci-e2e.sh#L62

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:54]: added doc I prefer to keep the generic name as it may do more than just creating bmh for example exporting some vars and to be consistent with other hooks

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:94]: if exported before it breaks creating the management because it uses v1beta1 templates

Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:94]: WIP comming soon :relaxed:

File: scripts/environment.sh
Comment (mboukhalfa) [reply scripts/environment.sh:64]: yeah for some reason when devenv install krew it will not be available in the path here ! I had to add it

File: scripts/environment.sh
Comment (mboukhalfa) [reply scripts/environment.sh:67]: capi test add scale up check after upgrade and for this we need additional bmh https://github.com/kubernetes-sigs/cluster-api/blob/4d8c88f24430d4fd15109e53fba57ad2acf5ddc8/test/e2e/clusterctl_upgrade.go#L393

File: test/e2e/config/e2e_conf.yaml
Comment (mboukhalfa) [reply test/e2e/config/e2e_conf.yaml:152]: hmm for now we are using what ever capi would define in their e2e config file but we can override that from the environment variable

File: scripts/environment.sh
Comment (mboukhalfa) [reply scripts/environment.sh]: THANKS

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go]: done

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:98]: done

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go]: fixed

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:140]: I still want to keep same generic names for call hooks added docs

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go]: done

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go]: done

File: test/e2e/upgrade_management_cluster_test.go
Comment (lentzi90) [reply test/e2e/upgrade_management_cluster_test.go:66]: Ok my point is not the outer function. It is clear that it just returns the inner function. My point is the inner function has no description for what it is doing and also no name so there is no way to understand it other than reading the code. It is not necessary to do it in this way. You can define the function separately and give it a name, then just pass it to the upgrade spec directly.
Like this:

```go
func myFancyPreWaitForClusterFunction(clusterProxy framework.ClusterProxy, clusterNamespace string, clusterName string) {
...
}

var _ = Describe("When testing cluster upgrade v1alpha5 > current [upgrade]", func() {
	capi_e2e.ClusterctlUpgradeSpec(ctx, func() capi_e2e.ClusterctlUpgradeSpecInput {
		return capi_e2e.ClusterctlUpgradeSpecInput{
			E2EConfig:                 e2eConfig,
			ClusterctlConfigPath:      clusterctlConfigPath,
			BootstrapClusterProxy:     bootstrapClusterProxy,
			ArtifactFolder:            artifactFolder,
			SkipCleanup:               true, // parameterize after merging https://github.com/kubernetes-sigs/cluster-api/pull/7373
			InitWithProvidersContract: "v1alpha4",
			InitWithBinary:            "https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.4.7/clusterctl-{OS}-{ARCH}",
			PreInit:                   getPreInitFunc(ctx),
			PreWaitForCluster:         myFancyPreWaitForClusterFunction,
			PreUpgrade:                getPreUpgrade(ctx),
			MgmtFlavor:                osType,
			WorkloadFlavor:            "ubuntu",
		}
	})
})
```

File: test/e2e/config/e2e_conf.yaml
Comment (mboukhalfa) [reply test/e2e/config/e2e_conf.yaml]: in fact this will be used in the in the capi upgrade test since we are passing our e2e config as input and and the test look for a var with this name although it has same value as `UPGRADED_K8S_VERSION`

**Review by lentzi90 — COMMENTED** [review]
Thanks for working on this!

Now I'm happy with how we pass the functions to the upgrade spec input. But I still want better descriptions and names for them. It is obvious that they are used for `Pre{Init|WaitForCluster|Upgrade}` since we pass them to the spec, but we need an explanation for what they are actually doing.

So for PreInit for example, ok we pass a function to this so it is obvious we want to do something before initializing. What are we doing? That is the question. On a high level maybe `SetupIronicBMO` could work?

**Comment by mboukhalfa** [comment]
> Thanks for working on this!
> 
> Now I'm happy with how we pass the functions to the upgrade spec input. But I still want better descriptions and names for them. It is obvious that they are used for `Pre{Init|WaitForCluster|Upgrade}` since we pass them to the spec, but we need an explanation for what they are actually doing.
> 
> So for PreInit for example, ok we pass a function to this so it is obvious we want to do something before initializing. What are we doing? That is the question. On a high level maybe `SetupIronicBMO` could work?

I think the current names are better to easily find which function runs in which step and as the outer function just decorate inner functions that have specific names I think that's better

**Review by lentzi90 — COMMENTED** [review]
I would still want more descriptive names for the `pre*` functions, but I can live with the descriptions. Let's see what others say.

/lgtm

**Comment by mboukhalfa** [comment]
> I would still want more descriptive names for the `pre*` functions, but I can live with the descriptions. Let's see what others say.
> 
> /lgtm

I guess the current way is more conform with CAPI since pre is common with all hooks and differentiate them from other functions

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:33]: linked PR is merged, can you do the parameterization now?

File: test/e2e/config/e2e_conf.yaml
Comment (furkatgofurov7) [reply test/e2e/config/e2e_conf.yaml]: Let's set it to `UPGRADED_K8S_VERSION` not to change it in two places in the future?

File: scripts/environment.sh
Comment (furkatgofurov7) [reply scripts/environment.sh]: I am not sure got this right

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:90]: how this is iterating through if you have only two/there BMHs here/below?

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:102]: why this is set at the end?

File: test/e2e/upgrade_management_cluster_test.go
Comment (furkatgofurov7) [inline test/e2e/upgrade_management_cluster_test.go:143]: same here, after all why do not use set them in the e2e conf file?

**Review by furkatgofurov7 — COMMENTED** [review]
This is good already, my second pass on it:

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:33]: We should wait for the new patch release

**Comment by furkatgofurov7** [comment]
> I guess the current way is more conform with CAPI since pre is common with all hooks and differentiate them from other functions

+1 to keep it sync with CAPI, unless we have a proper description for them

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:143]: this specific to the workload if we exported then before we will override the first management cluster

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:102]: we can move this to pre-init with the IP override it is the same at the end what is important is to export this after applying the first target cluster template

Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:102]: or even export is in the if block `if isBootstrapProxy {`

File: test/e2e/upgrade_management_cluster_test.go
Comment (mboukhalfa) [reply test/e2e/upgrade_management_cluster_test.go:90]: we know that file are ordered `secret_node_0.yaml, bmh_node_0.yaml, ...,secret_node_n.yaml, bmh_node_n.yaml..` so the first iteration is to apply the secret then the second to apply bmh for each node

**Comment by mboukhalfa** [comment]
/test-ubuntu-integration-main
/test-upgrade-e2e-main

@lentzi90  can you please re LGTM if the recent changes are ok for you

**Comment by lentzi90** [comment]
> I guess the current way is more conform with CAPI since pre is common with all hooks and differentiate them from other functions

I don't agree. CAPI only provide the spec. They don't even use these functions. To name our custom functions the same as the spec is not the same as conforming with CAPI. We use their spec and that is enough to make it conformant. What we name the input is up to us, and IMO should describe what it actually does, not "when it is called".
Just for the record :wink: 

But I'm still ok with merging if others are happy with this naming.
/lgtm

File: test/e2e/config/e2e_conf.yaml
Comment (furkatgofurov7) [inline test/e2e/config/e2e_conf.yaml:136]: @mboukhalfa what is the reason behind pinning it to this specific version? due to v1a5? please keep in mind https://github.com/metal3-io/cluster-api-provider-metal3/pull/755 is bumping it to 1.9.1

Comment (mboukhalfa) [reply test/e2e/config/e2e_conf.yaml:136]: yes this is the version used by CAPI `v0.4.8` I am not sure if we uplifted to a newer version what will happen after `clusterctl init ` anyway I prefer to merge this PR then #755 will be rebased and can use the version the more adequate

**Review by furkatgofurov7 — APPROVED** [review]

File: test/e2e/config/e2e_conf.yaml
Comment (furkatgofurov7) [reply test/e2e/config/e2e_conf.yaml:136]: 755 bumps it in Tiltfile, so this should be fine.

**Pull request merged**
